tfimm/architectures/lora/layers.py

```diff
@@ -111,14 +114,138 @@ def get_config(self):
+        bias_initializer="zeros",
+        kernel_regularizer=None,
+        bias_regularizer=None,
+        activity_regularizer=None,
+        kernel_constraint=None,
+        bias_constraint=None,
+        lora_rank: int = 4,
+        lora_alpha: float = 1,
+        **kwargs,
+    ):
+        super().__init__(
+            filters=filters,
+            kernel_size=kernel_size,
+            strides=strides,
+            padding=padding,
+            data_format=data_format,
+            dilation_rate=dilation_rate,
+            groups=groups,
+            activation=activation,
+            use_bias=use_bias,
+            kernel_initializer=kernel_initializer,
+            bias_initializer=bias_initializer,
+            kernel_regularizer=kernel_regularizer,
+            bias_regularizer=bias_regularizer,
+            activity_regularizer=activity_regularizer,
+            kernel_constraint=kernel_constraint,
+            bias_constraint=bias_constraint,
+            **kwargs,
+        )
+        self.lora_rank = lora_rank
+        self.lora_alpha = lora_alpha
+        self.scaling = lora_alpha / lora_rank
+
+        self.kernel_lora_a = None
+        self.kernel_lora_b = None
+
+        self.merged = False
+        self.full_rank_kernel = None
+
+    def build(self, input_shape):
+        super().build(input_shape)
+        kernel_height, kernel_width, in_channels, out_channels = self.kernel.shape
+        self.kernel_lora_a = self.add_weight(
+            "kernel_lora_a",
+            shape=(kernel_height, kernel_width, in_channels, self.lora_rank),
+            # For now, we are reusing the class default parameter. We could make this
+            # customisable. Note that we cannot simply use self.kernel_initializer here,
+            # because initializers should only be used once.
+            initializer="glorot_uniform",
+            regularizer=self.kernel_regularizer,
+            # We don't support constraints on the low-rank updates at the moment.
+            constraint=None,
+            dtype=self.dtype,
+            trainable=True,
+        )
+        self.kernel_lora_b = self.add_weight(
+            "kernel_lora_b",
+            shape=(kernel_height, kernel_width, self.lora_rank, out_channels),
+            initializer=tf.keras.initializers.Zeros(),
+            regularizer=self.kernel_regularizer,
+            constraint=None,
+            dtype=self.dtype,
+            trainable=True,
+        )
+
+    def call(self, x):
+        if self.merged:
+            return super().call(x)
+
+        full_kernel = self.kernel.value()  # Save the existing kernel
+
+        # Apply low-rank update to kernel
+        kernel_update = tf.matmul(self.kernel_lora_a, self.kernel_lora_b)
+        self.kernel.assign_add(self.scaling * kernel_update)
+        x = super().call(x)  # Re-use the call function of the Keras Cond2D layer
+
+        self.kernel.assign(full_kernel)  # Restore kernel again
+        return x
+
```

Comment:
the functions below could belong to a `LoRABaseLayer`, maybe something to do for the next added layer.

Reply:
Yes, its worth considering.